BrokerReprocessingTest.shouldCorrelateMessageAfterRestartIfPublishedBefore is flaky

**Summary**

- How often does the test fail?
  - at least 5 times so far
- Does it block your work?
  - no, but it's a little annoying
- Do we suspect that it is a real failure?
  - no, it's either just flaky or something with the new pipeline

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
io.camunda.zeebe.client.api.command.ClientStatusException: Unexpected error occurred during the request processing: No remote message handler registered for this message, subject command-api-1
	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.transformExecutionException(ZeebeClientFutureImpl.java:93)
	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:50)
	at io.camunda.zeebe.it.startup.BrokerReprocessingTest.startProcessInstance(BrokerReprocessingTest.java:593)
	at io.camunda.zeebe.it.startup.BrokerReprocessingTest.shouldCorrelateMessageAfterRestartIfPublishedBefore(BrokerReprocessingTest.java:539)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:258)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: INTERNAL: Unexpected error occurred during the request processing: No remote message handler registered for this message, subject command-api-1
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:48)
	... 19 more
Caused by: io.grpc.StatusRuntimeException: INTERNAL: Unexpected error occurred during the request processing: No remote message handler registered for this message, subject command-api-1
	at io.grpc.Status.asRuntimeException(Status.java:535)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:487)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:562)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:743)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:722)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	... 1 more
</pre>
</details>

--- a/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartition.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartition.java
@@ -246,9 +246,10 @@ public final class ZeebePartition extends Actor
                 t -> {
                   if (t != null) {
                     onInstallFailure(t);
+                  } else {
+                    onRecoveredInternal();
                   }
                 });
-            onRecoveredInternal();
           } else {
             onInstallFailure(error);
           }
@@ -269,9 +270,10 @@ public final class ZeebePartition extends Actor
                   // Compare with the current term in case a new role transition happened
                   if (t != null) {
                     onInstallFailure(t);
+                  } else {
+                    onRecoveredInternal();
                   }
                 });
-            onRecoveredInternal();
           } else {
             onInstallFailure(error);
           }

--- a/broker/src/test/java/io/camunda/zeebe/broker/test/EmbeddedBrokerRule.java
+++ b/broker/src/test/java/io/camunda/zeebe/broker/test/EmbeddedBrokerRule.java
@@ -26,9 +26,8 @@ import io.camunda.zeebe.broker.system.SystemContext;
 import io.camunda.zeebe.broker.system.configuration.BrokerCfg;
 import io.camunda.zeebe.engine.state.QueryService;
 import io.camunda.zeebe.gateway.impl.broker.BrokerClient;
-import io.camunda.zeebe.gateway.impl.broker.cluster.BrokerClusterState;
-import io.camunda.zeebe.gateway.impl.broker.cluster.BrokerTopologyManager;
 import io.camunda.zeebe.logstreams.log.LogStream;
+import io.camunda.zeebe.protocol.record.PartitionHealthStatus;
 import io.camunda.zeebe.test.util.TestConfigurationFactory;
 import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
 import io.camunda.zeebe.test.util.socket.SocketUtil;
@@ -249,9 +248,15 @@ public final class EmbeddedBrokerRule extends ExternalResource {
 
       waitUntil(
           () -> {
-            final BrokerTopologyManager topologyManager = brokerClient.getTopologyManager();
-            final BrokerClusterState topology = topologyManager.getTopology();
-            return topology != null && topology.getLeaderForPartition(1) >= 0;
+            final var topology = brokerClient.getTopologyManager().getTopology();
+            if (topology == null) {
+              return false;
+            }
+            final var leader = topology.getLeaderForPartition(1);
+            if (leader < 0) {
+              return false;
+            }
+            return topology.getPartitionHealth(leader, 1) == PartitionHealthStatus.HEALTHY;
           });
     }
 

--- a/broker/src/test/java/io/camunda/zeebe/broker/test/EmbeddedBrokerRule.java
+++ b/broker/src/test/java/io/camunda/zeebe/broker/test/EmbeddedBrokerRule.java
@@ -13,7 +13,6 @@ import static io.camunda.zeebe.broker.test.EmbeddedBrokerConfigurator.setCommand
 import static io.camunda.zeebe.broker.test.EmbeddedBrokerConfigurator.setGatewayApiPort;
 import static io.camunda.zeebe.broker.test.EmbeddedBrokerConfigurator.setGatewayClusterPort;
 import static io.camunda.zeebe.broker.test.EmbeddedBrokerConfigurator.setInternalApiPort;
-import static io.camunda.zeebe.test.util.TestUtil.waitUntil;
 
 import io.atomix.cluster.AtomixCluster;
 import io.camunda.zeebe.broker.Broker;
@@ -21,14 +20,13 @@ import io.camunda.zeebe.broker.PartitionListener;
 import io.camunda.zeebe.broker.SpringBrokerBridge;
 import io.camunda.zeebe.broker.TestLoggers;
 import io.camunda.zeebe.broker.clustering.ClusterServices;
-import io.camunda.zeebe.broker.system.EmbeddedGatewayService;
 import io.camunda.zeebe.broker.system.SystemContext;
 import io.camunda.zeebe.broker.system.configuration.BrokerCfg;
+import io.camunda.zeebe.client.ZeebeClient;
 import io.camunda.zeebe.engine.state.QueryService;
-import io.camunda.zeebe.gateway.impl.broker.BrokerClient;
 import io.camunda.zeebe.logstreams.log.LogStream;
-import io.camunda.zeebe.protocol.record.PartitionHealthStatus;
 import io.camunda.zeebe.test.util.TestConfigurationFactory;
+import io.camunda.zeebe.test.util.asserts.TopologyAssert;
 import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
 import io.camunda.zeebe.test.util.socket.SocketUtil;
 import io.camunda.zeebe.util.FileUtil;
@@ -36,6 +34,7 @@ import io.camunda.zeebe.util.allocation.DirectBufferAllocator;
 import io.camunda.zeebe.util.sched.clock.ControlledActorClock;
 import io.camunda.zeebe.util.sched.future.ActorFuture;
 import io.camunda.zeebe.util.sched.future.CompletableActorFuture;
+import io.netty.util.NetUtil;
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
@@ -49,6 +48,7 @@ import java.util.function.Consumer;
 import java.util.function.Supplier;
 import org.agrona.LangUtil;
 import org.assertj.core.util.Files;
+import org.awaitility.Awaitility;
 import org.junit.rules.ExternalResource;
 import org.junit.runner.Description;
 import org.junit.runners.model.Statement;
@@ -241,23 +241,23 @@ public final class EmbeddedBrokerRule extends ExternalResource {
       Thread.currentThread().interrupt();
     }
 
-    final EmbeddedGatewayService embeddedGatewayService =
-        broker.getBrokerContext().getEmbeddedGatewayService();
-    if (embeddedGatewayService != null) {
-      final BrokerClient brokerClient = embeddedGatewayService.get().getBrokerClient();
-
-      waitUntil(
-          () -> {
-            final var topology = brokerClient.getTopologyManager().getTopology();
-            if (topology == null) {
-              return false;
-            }
-            final var leader = topology.getLeaderForPartition(1);
-            if (leader < 0) {
-              return false;
-            }
-            return topology.getPartitionHealth(leader, 1) == PartitionHealthStatus.HEALTHY;
-          });
+    if (brokerCfg.getGateway().isEnable()) {
+      try (final var client =
+          ZeebeClient.newClientBuilder()
+              .gatewayAddress(NetUtil.toSocketAddressString(getGatewayAddress()))
+              .usePlaintext()
+              .build()) {
+        Awaitility.await("until we have a complete topology")
+            .untilAsserted(
+                () -> {
+                  final var topology = client.newTopologyRequest().send().join();
+                  TopologyAssert.assertThat(topology)
+                      .isComplete(
+                          brokerCfg.getCluster().getClusterSize(),
+                          brokerCfg.getCluster().getPartitionsCount())
+                      .isHealthy();
+                });
+      }
     }
 
     dataDirectory = broker.getSystemContext().getBrokerConfiguration().getData().getDirectory();

--- a/test-util/src/main/java/io/camunda/zeebe/test/util/asserts/TopologyAssert.java
+++ b/test-util/src/main/java/io/camunda/zeebe/test/util/asserts/TopologyAssert.java
@@ -8,6 +8,7 @@
 package io.camunda.zeebe.test.util.asserts;
 
 import io.camunda.zeebe.client.api.response.BrokerInfo;
+import io.camunda.zeebe.client.api.response.PartitionBrokerHealth;
 import io.camunda.zeebe.client.api.response.PartitionInfo;
 import io.camunda.zeebe.client.api.response.Topology;
 import java.util.List;
@@ -100,6 +101,18 @@ public final class TopologyAssert extends AbstractObjectAssert<TopologyAssert, T
           count, actual.getBrokers());
     }
 
+    return myself;
+  }
+
+  public TopologyAssert isHealthy() {
+    isNotNull();
+    final var partitions =
+        actual.getBrokers().stream()
+            .flatMap(brokerInfo -> brokerInfo.getPartitions().stream())
+            .collect(Collectors.toList());
+    newListAssertInstance(partitions)
+        .as("all partitions are healthy")
+        .allMatch(partition -> partition.getHealth() == PartitionBrokerHealth.HEALTHY);
     return myself;
   }
 }

--- a/test-util/src/main/java/io/camunda/zeebe/test/util/asserts/TopologyAssert.java
+++ b/test-util/src/main/java/io/camunda/zeebe/test/util/asserts/TopologyAssert.java
@@ -26,7 +26,7 @@ public final class TopologyAssert extends AbstractObjectAssert<TopologyAssert, T
     return new TopologyAssert(actual);
   }
 
-  public final TopologyAssert isComplete(final int clusterSize, final int partitionCount) {
+  public TopologyAssert isComplete(final int clusterSize, final int partitionCount) {
     isNotNull();
 
     final List<BrokerInfo> brokers = actual.getBrokers();
@@ -68,7 +68,7 @@ public final class TopologyAssert extends AbstractObjectAssert<TopologyAssert, T
     return myself;
   }
 
-  public final TopologyAssert doesNotContainBroker(final int nodeId) {
+  public TopologyAssert doesNotContainBroker(final int nodeId) {
     isNotNull();
 
     final List<Integer> brokers =
@@ -82,7 +82,7 @@ public final class TopologyAssert extends AbstractObjectAssert<TopologyAssert, T
     return myself;
   }
 
-  public final TopologyAssert hasBrokerSatisfying(final Consumer<BrokerInfo> condition) {
+  public TopologyAssert hasBrokerSatisfying(final Consumer<BrokerInfo> condition) {
     isNotNull();
 
     final List<BrokerInfo> brokers = actual.getBrokers();
@@ -91,7 +91,7 @@ public final class TopologyAssert extends AbstractObjectAssert<TopologyAssert, T
     return myself;
   }
 
-  public final TopologyAssert hasBrokersCount(final int count) {
+  public TopologyAssert hasBrokersCount(final int count) {
     isNotNull();
 
     if (actual.getBrokers().size() != count) {

